File: ios/Runner/SalesforceMethodChannelHandler.swift
import Flutter
import UIKit
import os.log

/// Handles Flutter method channel communication for Salesforce In-App Chat
class SalesforceMethodChannelHandler: SalesforceAuthTokenProvider {

    static let shared = SalesforceMethodChannelHandler()

    private let channelName = "com.newyorklife.mynyl.mobile/salesforce_chat"
    private var channel: FlutterMethodChannel?

    private init() {}

    /// Setup method channel for Flutter communication
    func setup(with engine: FlutterEngine) {
        channel = FlutterMethodChannel(
            name: channelName,
            binaryMessenger: engine.binaryMessenger
        )

        channel?.setMethodCallHandler { [weak self] (call, result) in
            guard let self = self else {
                result(
                    FlutterError(
                        code: "HANDLER_ERROR", message: "Handler deallocated", details: nil))
                return
            }
            self.handleMethodCall(call, result: result)
        }

        // Set the channel reference in SalesforceMessagingManager so it can send events
        SalesforceMessagingManager.shared.setEventChannel(channel)
        SalesforceMessagingManager.shared.setAuthTokenProvider(self)

        print("Salesforce Method Channel Handler setup complete for channel: \(channelName)")
    }

    /// Invoke a method on Flutter side
    func invokeFlutterMethod(_ method: String, arguments: Any? = nil) {
        channel?.invokeMethod(method, arguments: arguments) { _ in }
    }

    /// Handle incoming method calls from Flutter
    private func handleMethodCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard
            let window = UIApplication.shared.connectedScenes
                .compactMap({ $0 as? UIWindowScene })
                .flatMap({ $0.windows })
                .first(where: { $0.isKeyWindow }),
            let rootViewController = window.rootViewController
        else {
            result(
                FlutterError(
                    code: "NO_VIEW_CONTROLLER", message: "Root view controller not found",
                    details: nil))
            return
        }

        switch call.method {
        case "openChat":
            handleOpenChat(call, viewController: rootViewController, result: result)

        case "clearConversation":
            let args = call.arguments as? [String: Any]
            let deletePersistedConversationId =
                args?["deletePersistedConversationId"] as? Bool ?? true
            SalesforceMessagingManager.shared.clearConversation(
                deletePersistedConversationId: deletePersistedConversationId)
            result(true)

        case "closeConversation":
            SalesforceMessagingManager.shared.closeConversation { error in
                if let error = error {
                    result(
                        FlutterError(
                            code: "CLOSE_ERROR", message: error.localizedDescription, details: nil))
                } else {
                    result(true)
                }
            }

        case "minimizeChat":
            SalesforceMessagingManager.shared.minimizeChat()
            result(true)

        case "getConversationId":
            let conversationId = SalesforceMessagingManager.shared.getCurrentConversationId()
            result(conversationId?.uuidString)

        case "revokeToken":
            SalesforceMessagingManager.shared.revokeToken()
            result(true)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    func onGetToken(completion: @escaping (String?) -> Void) {
        requestTokenFromFlutter(method: "getToken", completion: completion)
    }

    func onRefreshToken(completion: @escaping (String?) -> Void) {
        requestTokenFromFlutter(method: "refreshToken", completion: completion)
    }

    private func requestTokenFromFlutter(method: String, completion: @escaping (String?) -> Void) {
        guard let channel = channel else {
            print("[SalesforceMethodChannelHandler] ERROR: Channel unavailable for \(method)")
            completion(nil)
            return
        }

        func requestToken(attempt: Int) {
            channel.invokeMethod(method, arguments: nil) { response in
                if let token = response as? String, !token.isEmpty {
                    print(
                        "[SalesforceMethodChannelHandler] \(method) token response length: \(token.count) (attempt=\(attempt))"
                    )
                    completion(token)
                    return
                }

                if attempt == 1 {
                    print("[SalesforceMethodChannelHandler] \(method) empty token on attempt=1, retrying")
                    DispatchQueue.main.asyncAfter(deadline: .now() + 0.25) {
                        requestToken(attempt: 2)
                    }
                    return
                }

                completion(nil)
            }
        }

        DispatchQueue.main.async {
            requestToken(attempt: 1)
        }
    }

    /// Handle openChat method call
    private func handleOpenChat(
        _ call: FlutterMethodCall, viewController: UIViewController, result: @escaping FlutterResult
    ) {
        let args = call.arguments as? [String: Any]

        do {
            guard let serviceApiUrl = args?["serviceApiUrl"] as? String,
                let orgId = args?["orgId"] as? String,
                let deploymentName = args?["deploymentName"] as? String
            else {
                result(
                    FlutterError(
                        code: "INVALID_ARGS", message: "Missing configuration parameters",
                        details: nil))
                return
            }

            let chatAssistantTitle = args?["chatAssistantTitle"] as? String
            let chatAssistantTooltipMessage = args?["chatAssistantTooltipMessage"] as? String

            let hasPreChatValues = args?["hasPreChatValues"] as? Bool ?? false
            let clientId = args?["clientId"] as? String
            let policyNumber = args?["policyNumber"] as? String
            let reason = args?["reason"] as? String
            let timeZoneOffset = args?["timeZoneOffset"] as? String
            let hasAllPreChatValues =
                !(clientId?.isEmpty ?? true)
                && !(policyNumber?.isEmpty ?? true)
                && !(reason?.isEmpty ?? true)
                && !(timeZoneOffset?.isEmpty ?? true)

            os_log(
                "[SalesforceMethodChannelHandler] openChat manual mode hasPreChatValues=\(hasPreChatValues)"
            )

            if hasPreChatValues && !hasAllPreChatValues {
                result(
                    FlutterError(
                        code: "INVALID_ARGS",
                        message:
                            "hasPreChatValues is true but one or more pre-chat fields are missing",
                        details: [
                            "clientId": clientId != nil,
                            "policyNumber": policyNumber != nil,
                            "reason": reason != nil,
                            "timeZoneOffset": timeZoneOffset != nil,
                        ]
                    ))
                return
            }

            if hasAllPreChatValues,
                let clientId = clientId,
                let policyNumber = policyNumber,
                let reason = reason,
                let timeZoneOffset = timeZoneOffset
            {
                os_log(
                    "[SalesforceMethodChannelHandler] pre-chat payload lengths clientId=\(clientId.count) policyNumber=\(policyNumber.count) reason=\(reason.count) timeZoneOffset=\(timeZoneOffset.count)"
                )
                try SalesforceMessagingManager.shared.openChat(
                    from: viewController,
                    serviceApiUrl: serviceApiUrl,
                    orgId: orgId,
                    deploymentName: deploymentName,
                    clientId: clientId,
                    policyNumber: policyNumber,
                    reason: reason,
                    timeZoneOffset: timeZoneOffset,
                    chatAssistantTitle: chatAssistantTitle,
                    chatAssistantTooltipMessage: chatAssistantTooltipMessage
                )
            } else {
                try SalesforceMessagingManager.shared.resumeChat(
                    from: viewController,
                    serviceApiUrl: serviceApiUrl,
                    orgId: orgId,
                    deploymentName: deploymentName,
                    chatAssistantTitle: chatAssistantTitle,
                    chatAssistantTooltipMessage: chatAssistantTooltipMessage
                )
            }
            invokeFlutterMethod("onChatOpened")
            result(true)
        } catch {
            result(
                FlutterError(code: "CHAT_ERROR", message: error.localizedDescription, details: nil))
        }
    }
}

File: android/app/src/main/kotlin/com/example/newproject/SalesforceMethodChannelHandler.kt
package com.newyorklife.mynyl.mobile

import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext
import kotlin.coroutines.resume

/**
 * Handles Flutter MethodChannel communication for Salesforce In-App Messaging.
 */
class SalesforceMethodChannelHandler(
    private val salesforceManager: SalesforceMessagingManager
) : SalesforceAuthTokenProvider {

    companion object {
        private const val CHANNEL = "com.newyorklife.mynyl.mobile/salesforce_chat"
        private var methodChannel: MethodChannel? = null

        fun sendEventToFlutter(event: String) {
            when (event) {
                "opened" -> methodChannel?.invokeMethod("onChatOpened", null)
                "minimized" -> methodChannel?.invokeMethod("onChatMinimized", null)
                "closed" -> methodChannel?.invokeMethod("onChatClosed", null)
                "session_ended" -> methodChannel?.invokeMethod("onSessionEnded", null)
            }
        }
    }

    private suspend fun fetchToken(method: String): String = withContext(Dispatchers.Main) {
        suspendCancellableCoroutine { continuation ->
            methodChannel?.invokeMethod(method, null, object : MethodChannel.Result {
                override fun success(p0: Any?) {
                    when (p0) {
                        is String -> continuation.resume(p0)
                    }

                }

                override fun error(p0: String, p1: String?, p2: Any?) {
                    TODO("Not yet implemented")
                }

                override fun notImplemented() {
                    TODO("Not yet implemented")
                }

            })

        }
    }

    override suspend fun onGetToken(): String = fetchToken("getToken")

    override suspend fun onRefreshToken(): String = fetchToken("refreshToken")

    fun register(flutterEngine: FlutterEngine) {
        methodChannel = MethodChannel(flutterEngine.dartExecutor.binaryMessenger, CHANNEL)
        methodChannel?.setMethodCallHandler { call, result ->
            when (call.method) {
                "openChat" -> {
                    try {
                        val useConfigFile = call.argument<Boolean>("useConfigFile") ?: true
                        val persistConversation = call.argument<Boolean>("persistConversation") ?: true

                        if (useConfigFile) {
                            salesforceManager.openChatWithConfigFile(persistConversation)
                        } else {
                            val serviceApiUrl = call.argument<String>("serviceApiUrl")
                            val orgId = call.argument<String>("orgId")
                            val deploymentName = call.argument<String>("deploymentName")
                            val hasPreChatValues = call.argument<Boolean>("hasPreChatValues") ?: false

                            if (serviceApiUrl != null && orgId != null && deploymentName != null) {
                                if (hasPreChatValues) {
                                    val clientId = call.argument<String>("clientId")
                                    val policyNumber = call.argument<String>("policyNumber")
                                    val reason = call.argument<String>("reason")
                                    val timeZoneOffset = call.argument<String>("timeZoneOffset")

                                    if (clientId != null && policyNumber != null && reason != null && timeZoneOffset != null) {
                                        salesforceManager.openChatManualWithPreChatValues(
                                            persistConversation,
                                            serviceApiUrl,
                                            orgId,
                                            deploymentName,
                                            clientId,
                                            policyNumber,
                                            reason,
                                            timeZoneOffset
                                        )
                                    }
                                } else {
                                    salesforceManager.openChatManual(
                                        persistConversation,
                                        serviceApiUrl,
                                        orgId,
                                        deploymentName
                                    )
                                }
                            } else {
                                result.error("INVALID_ARGS", "Manual config requires serviceApiUrl, orgId, and deploymentName", null)
                                return@setMethodCallHandler
                            }
                        }
                        sendEventToFlutter("opened")
                        result.success(true)
                    } catch (e: Exception) {
                        result.error("CHAT_ERROR", e.message, null)
                    }
                }
                "clearConversation" -> {
                    try {
                        salesforceManager.clearConversation()
                        result.success(true)
                    } catch (e: Exception) {
                        result.error("CLEAR_ERROR", e.message, null)
                    }
                }
                "closeConversation" -> {
                    try {
                        val success = salesforceManager.closeConversation()
                        if (success) {
                            sendEventToFlutter("session_ended")
                        }
                        result.success(success)
                    } catch (e: Exception) {
                        result.error("CLOSE_ERROR", e.message, null)
                    }
                }
                "startNewConversation" -> {
                    try {
                        salesforceManager.startNewConversation()
                        sendEventToFlutter("opened")
                        result.success(true)
                    } catch (e: Exception) {
                        result.error("NEW_CONVERSATION_ERROR", e.message, null)
                    }
                }
                "minimizeChat" -> {
                    // Handled by native chat UI controls; Flutter can listen to callbacks.
                    result.success(true)
                }
                "getConversationId" -> {
                    val conversationId = salesforceManager.getCurrentConversationId()
                    result.success(conversationId?.toString())
                }
                "revokeToken" -> {
                    try {
                        salesforceManager.revokeToken()
                        result.success(true)
                    } catch (e: Exception) {
                        result.error("REVOKE_ERROR", e.message, null)
                    }
                }
                else -> {
                    result.notImplemented()
                }
            }
        }
    }
}
